### PR TITLE
Update block/chain segment processing

### PR DESCRIPTION
**Pull request by wemeetagain**
- Add `ChainSegmentError` - mirrors `BlockError` but includes # of `importedBlocks` before error
- Add `processChainSegment` - mirrors `processBlock` but with epoch-by-epoch signature verification
- Move batch signature verification into `processBlock` / `processChainSegment` rather than lower-level `runStateTransition`
- Move check for parent parent from `validateBlock` into `processBlock` / `processChainSegment`
- Update `processChainSegmentJob` to do filtering of irrelevant blocks & use `processChainSegment`

### CODE REVIEW

File: packages/lodestar/src/chain/blocks/processor.ts [hunk 1]
```diff
@@ -72,20 +74,85 @@ export async function processBlockJob(modules: BlockProcessorModules, job: IBloc
...
-    // above functions only throw BlockError
-    modules.emitter.emit(ChainEvent.errorBlock, e);
-    throw e;
-  }
+export async function processChainSegmentJob(modules: BlockProcessorModules, job: IChainSegmentJob): Promise<void> {
+  const blocks = job.signedBlocks;
+
+  // Validate and filter out irrelevant blocks
+  const filteredChainSegment: SignedBeaconBlock[] = [];
+  blocks.forEach(async (block, i) => {
```
Comment (dapplion) [inline packages/lodestar/src/chain/blocks/processor.ts]: Would recommend to use
```ts
for (const [i, block] of blocks.entries()) {
  //
}
```
instead

File: packages/lodestar/src/chain/blocks/validate.ts [hunk 2]
```diff
@@ -6,86 +6,63 @@ import {IBlockJob} from "../interface";
 import {IBeaconClock} from "../clock";
 import {BlockError, BlockErrorCode} from "../errors";
 
-export async function validateBlocks({
+export async function validateBlock({
```
Comment (dapplion) [inline packages/lodestar/src/chain/blocks/validate.ts]: No need for this function to be async anymore, there are no awaits inside

File: packages/lodestar/src/chain/blocks/processor.ts (hunk 1 above)
Comment (dapplion) [reply packages/lodestar/src/chain/blocks/processor.ts]: Also `validateBlock` should not be async, thus this function neither. See https://github.com/ChainSafe/lodestar/pull/1966/files#r559793686

File: packages/lodestar/src/chain/blocks/process.ts [hunk 3]
```diff
@@ -1,48 +1,189 @@
...
+      throw new ChainSegmentError({
+        code: BlockErrorCode.PARENT_UNKNOWN,
+        parentRoot: firstBlock.message.parentRoot.valueOf() as Uint8Array,
+        job,
+        importedBlocks,
+      });
+    }
+    const startEpoch = computeEpochAtSlot(config, firstBlock.message.slot);
+
+    const findLastIndex = <T>(array: T[], predicate: (value: T) => boolean): number => {
```
Comment (dapplion) [inline packages/lodestar/src/chain/blocks/process.ts]: This should go to src/util/array and be unit tested

**Review by dapplion — APPROVED** [review]

**Review by mpetrunic — APPROVED** [review]

**Pull request merged**